<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Kraken : A universal genomic coordinate translator for comparative genomics">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Kraken</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/nedaz/kraken">View on GitHub</a>

          <h1 id="project_title">Kraken</h1>
          <h2 id="project_tagline">A universal genomic coordinate translator for comparative genomics</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/nedaz/kraken/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/nedaz/kraken/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="methods" class="anchor" href="#methods" aria-hidden="true"><span class="octicon octicon-link"></span></a>Methods</h3>

<p><a href="http://www.biomedcentral.com/1471-2105/15/227?utm_source=dlvr.it&amp;utm_medium=tumblr">Article - BMC Bioinformatics</a></p>

<hr>

<h3>
<a id="overview" class="anchor" href="#overview" aria-hidden="true"><span class="octicon octicon-link"></span></a>Overview</h3>

<p>Kraken is a computational genome coordinate translator that facilitates cross-species comparisons, distinguishes orthologs from paralogs, and does not require costly all-to-all whole genome mappings.</p>

<hr>

<h3>
<a id="getting-started" class="anchor" href="#getting-started" aria-hidden="true"><span class="octicon octicon-link"></span></a>Getting started</h3>

<p>Download the package, unzip and run make from the main directory.  Use gcc version 4.2 or higher for this. You can start by running the code with the data provided in the samples directory. This is explained below.</p>

<hr>

<hr>

<p><em><strong>Example</strong></em>
The following example performs a mapping of items in the GTF input file "dmel.gtf" from specie "dmel" to "dyak". The data for this example can be found in the sample folder. The output is written to mapped.gtf as default. You can run the command:
../runKraken  -c dere_dyak_dmel.config -s dmel.gtf -S dmel -T dyak</p>

<hr>

<p>For more information, see the software guide located in the repository (doc/softwareGuide.pdf)</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Kraken maintained by <a href="https://github.com/nedaz">nedaz</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
